src/nginx/module.cc

```diff
@@ -42,45 +42,70 @@
 #include "src/nginx/util.h"
 #include "src/nginx/version.h"
 
-using ::google::protobuf::util::error::Code;
 using ::google::api_manager::utils::Status;
+using ::google::protobuf::util::error::Code;
 
 namespace google {
 namespace api_manager {
 namespace nginx {
 
 namespace {
 
-// Updates r->uri and r->unparsed_uri, based on api basepath math.
-// Then returns ApiManager::ApiBasepathRewriteAction.
-// If the return value is ApiManager::ApiBasepathRewriteAction::REJECT,
-// then nginx should response status 404 NotFound
-ApiManager::ApiBasepathRewriteAction ngx_esp_rewrite_uri(
-    ngx_http_request_t *r, ngx_esp_loc_conf_t *lc) {
-  std::string unparsed_uri;
-  ApiManager::ApiBasepathRewriteAction unparsed_uri_action =
-      lc->esp->ReWriteURL(ngx_str_to_std(r->unparsed_uri), &unparsed_uri);
+// Internal debugging header
+std::string kXEndpointsDebugUrlRewrite = "x-endpoints-debug-url-rewrite";
+
+// decode encoded uri
+void url_decode(const std::string &uri, std::string &decoded) {
+  decoded.reserve(uri.length());
+  for (std::size_t i = 0; i < uri.length(); i++) {
+    if (int(uri[i]) == 37) {
+      int ii;
+      sscanf(uri.substr(i + 1, 2).c_str(), "%x", &ii);
+      char ch = static_cast<char>(ii);
+      decoded.append(std::string(1, ch));
```

Comment:
does std::string have append for a char, ?

Reply:
Replaced with decoded.append(1, ch)